doc: New developer getting started guide

dbkinder opened the pull request:
As presented to the TSC, Zephyr's out-of-box experience for new
developers is, well, complicated.  A number of suggestions were
presented including simplifying the getting started material to present
a straight-forward path through the setup and installation steps through
to getting a sample application built, flashed, and running.

This PR is a work-in-progress towards addressing this OOB experience
with a minimal-distractions version of the GSG. Alternatives, warnings,
and material that could lead the developer astray were moved to
alternative/advanced instruction documents (based on the previous
separate Linux/macOS/Windows setup guides) and a new "Beyond the GSG"
document.

We do take advantage of a sphinx-tabs extension for synchronized tabs to
present OS-specific instructions: clicking on one tab will display all
same-named tabs throughout the doc.

We hope (and will continue evaluating) that this new GSG gets developers
set up quickly and then we can send them along to other documents to
continue learning about Zephyr and trying other sample apps.

Appreciate your feedback.

Signed-off-by: David B. Kinder <david.b.kinder@intel.com>

dbkinder committed:
doc: New developer getting started guide

As presented to the TSC, Zephyr's out-of-box experience for new
developers is, well, complicated.  A number of suggestions were
presented including simplifying the getting started material to present
a straight-forward path through the setup and installation steps through
to getting a sample application built, flashed, and running.

This PR is a work-in-progress towards addressing this OOB experience
with a minimal-distractions version of the GSG. Alternatives, warnings,
and material that could lead the developer astray were moved to
alternative/advanced instruction documents (based on the previous
separate Linux/macOS/Windows setup guides) and a new "Beyond the GSG"
document.

We do take advantage of a sphinx-tabs extension for synchronized tabs to
present OS-specific instructions: clicking on one tab will display all
same-named tabs throughout the doc.

We hope (and will continue evaluating) that this new GSG gets developers
set up quickly and then we can send them along to other documents to
continue learning about Zephyr and trying other sample apps.

Thanks for all your previous feedback that I've worked
into this new version.

Signed-off-by: David B. Kinder <david.b.kinder@intel.com>

mbolivar reviewed (dismissed):
I think this is great work which obviously reflects a lot of serious research into how our documentation plays out in the real world. Thanks a lot for what I'm sure has been a dedicated and involved (and hopefully not entirely thankless) effort behind the scenes. I do have some initial feedback on the particulars, but I agree wholeheartedly with the general direction.

aescolar commented on doc/getting_started/index.rst:
```diff
@@ -3,319 +3,372 @@
+.. tabs::
 
-.. _python-pip:
+   .. group-tab:: Linux
```
This instructions being only for Ubuntu, I wonder if it would be good to make it a bit more discoverable for other distros. Maybe a link to docs/getting_started/installation_linux.html, maybe more tabs(?)

marc-hb replied:
There is no such thing as "Linux". Linux runs my smartphone and my wifi router and I doubt they will ever be able to build Zephyr. Please rename the "Linux" tab to "Ubuntu 12.34" and mention that other Debian-based distros should work the same for the most part - but even there your mileage may vary. 

I've been building Zephyr on Fedora with practically no issue *except* for the main one: manually "translating" the Getting Started Guide. It's fine, I'm used to it.

carlescufi replied:
I agree, it just doesn't work to write "Linux" here and then have instructions for Ubuntu. It will be confusing for users. Either write "Ubuntu" and then also add a link to the info for the rest of Linux derivatives or change this to contain the instructions for all distros.

aescolar reviewed (commented):
Thanks David, just some thoughts after a quick look

marc-hb commented on doc/getting_started/index.rst:
```diff
@@ -3,319 +3,372 @@
+         Subsystem for Linux (WSL). Using WSL will also let you to use
+         the Zephyr SDK. `Install the Windows Subsystem for Linux (WSL)
+         <https://msdn.microsoft.com/en-us/commandline/wsl/install_guide>`__, and
+         then follow the Linux (Ubuntu) Getting Started instructions.
```
I suspect the answer is "it depends". As a coincidence I'm using a USB-serial adapter from WSL right now and it Just Works.

carlescufi replied:
I think it is OK to have it but I would add a warning that the recommended option is to use `cmd.exe`

marc-hb replied:
I don't use Windows often but when I do I always use PowerShell. It's not perfect (nothing is) but it's orders of magnitude more usable than `cmd.exe`... Any particular reason for keeping Windows users stuck in the prehistory of computing? Because Windows users don't care, otherwise they'd be using Linux? :-P

dbkinder replied:
I followed the instructions as written for Linux/Ubuntu on WSL and made it all the way through without a hitch, until getting to "west flash" where it failed both using  just using "west flash" or using sudo:
```
~/zephyrproject/zephyr$ west flash
-- west flash: rebuilding
ninja: no work to do.
-- west flash: using runner pyocd
-- runners.pyocd: Flashing file: /home/dbkinder/zephyrproject/zephyr/build/zephyr/zephyr.hex
No connected debug probes
0014885:CRITICAL:__main__:uncaught exception:
Traceback (most recent call last):
  File "/home/dbkinder/.local/lib/python3.6/site-packages/pyocd/__main__.py", line 338, in run
    self._COMMANDS[self._args.cmd](self)
  File "/home/dbkinder/.local/lib/python3.6/site-packages/pyocd/__main__.py", line 468, in do_flash
    with session:
  File "/home/dbkinder/.local/lib/python3.6/site-packages/pyocd/core/session.py", line 284, in __enter__
    assert self._probe is not None
AssertionError
ERROR: command exited with status 1: pyocd flash -e sector -t nrf52 -f 4000000 /home/dbkinder/zephyrproject/zephyr/build/zephyr/zephyr.hex
```

@MaureenHelm comments on the [dev mailing list](https://lists.zephyrproject.org/g/users/topic/31871754#1496) that you "unfortunately cannot use pyocd in the Windows 10 Linux subsystem. WSL has limited support for USB", so we may be stuck. You can go so far using WSL, then need to switch to native windows tools for flashing/debugging?  That sounds like a messy solution vs. just using board/architecture-specific native windows tools (and not using the Zephyr SDK).

marc-hb replied:
> @marc-hb Have you flashed a board from WSL? We should talk.

No, I was referring to another, (simpler?) and working USB device. This device doesn't require using `root` BTW.

I suggest restricting mentions of WSL to the "advanced" page(s) with a short caveat that pyocd doesn't work last time someone tried (WSL is evolving very fast and running an actual Linux kernel "soon")

marc-hb commented on doc/getting_started/index.rst:
```diff
@@ -3,319 +3,372 @@
 
-Follow an operating system specific guide, then come back to this page.
+Next, use your OS package manager to install required OS-provided
+support tools.  For example, Python 3 and its package manager, pip, are
```
Drop "For example" as the entire purpose of pip is to... circumvent the OS package manager and provide much newer Python packages.

marc-hb commented on doc/getting_started/index.rst:
```diff
@@ -3,319 +3,372 @@
 
-You can run ``./build/zephyr/zephyr.exe --help`` to get a list of available
-options.
+         sudo -E PATH=$PATH sh -c "west flash"
```
> But if you add sudo, the newly generated files can be owned by root. 

Building as root is a big no-no. Especially not after pip has downloaded a bunch of random packages from the Internet, half of them users don't need (I understand different users need the other half)

> I think we cannot paper over the inherent complexity of device permissions on linux here, and ought to add information about necessary udev rules on supported platforms to the documentation as a troubleshooting step if this doesn't work.

Unfortunately agreed 100%. There's a lot that good automation and good documentation can achieve, but they can solve neither world hunger nor the complicated and fragmented state of udev security across Linux distributions. The best that can be done here is:
1. Make sure Ubuntu 12.34 works
2. For the 1000s of other Linux distributions, defer. Provide generic _test_ instructions or even better: generic test code that prints: "You don't seem to have permission to access to /dev/whatever. Please check the udev part of your Linux distribution"

BTW it would _really_ be nice for chocolatey to run without the Adminstrator account. I digress.

mbolivar replied:
I stand by the discussion from earlier. For whatever it's worth, flashing as root in the GSG is not OK with me.

If your flash tool installer doesn't set up udev rules properly, its installer is broken.

(Nordics' tools installed via .deb do take care of this properly, e.g.)

mbolivar replied:
I'm really, really not OK with flashing as root. I don't like that this was marked resolved without any changes.

dbkinder replied:
@mbolivar Do you have a counter proposal that would generically work for some set of supported boards?  I'm trying to find a simple path for a new developer to get a first sample app running on a board.  I tried saying using sudo is not the proper way, but we'll use it for this quick start. Let's talk offline and figure this out.

marc-hb replied:
Ignoring the security disaster for a moment, a build directory that can't be cleaned because it has a random number of files owned by root is everything but new user friendly!

> Do you have a counter proposal that would generically work for some set of supported boards?

A generic solution that works across a wide range of boards and Linux distros is probably impossible. On the other hand, a single udev rule that works reliably on one specific Ubuntu version for a reasonable number of boards shouldn't be rocket science.

Breaking news: security is rarely user-friendly.

dbkinder replied:
Talking with the bug triage folks, they recommend removing the ``sudo`` and just do ``west flash`` and add a troubleshooting note about udev rules.  As you mention, security isn't free.

mbolivar replied:
>  Talking with the bug triage folks, they recommend removing the sudo and just do west flash and add a troubleshooting note about udev rules. As you mention, security isn't free.

That's reasonable. Thanks.

mbolivar replied:
> a build directory that can't be cleaned because it has a random number of files owned by root is everything but new user friendly!

^^ also, it's not just a security thing. This is also an issue.

marc-hb commented on doc/getting_started/index.rst:
```diff
@@ -3,319 +3,372 @@
+         The ``--user`` option puts installed packages into your
+         ``~/.local/bin folder``.  It's very important you permanently update the
+         PATH environment variable with the location of these packages (in your
+         .bashrc profile, as shown).
```
Please avoid the word "profile" when referring to `.bashrc` considering the _alternatives_ to `.bashrc` are named `.profile`, `.bash_profile` and what not.

Has the .bashrc command above been fully tested when connecting remotely/graphically/scriptically/using zsh... ? (hi @carlescufi!)

Linux environment variables are a non-portable disaster: https://superuser.com/questions/183870/difference-between-bashrc-and-bash-profile/1452523#1452523
Yet another reason to rename the "Linux" tab in this page to "Ubuntu 12.34".

ru-fu replied:
I misread this sentence to mean that you should regularly update the PATH environment variable. Maybe rephrase to something like "Adding the PATH specification to your .bashrc file ensures that the change is permanent."

marc-hb commented on doc/getting_started/index.rst:
```diff
@@ -3,319 +3,372 @@
+         cd ~
+         west init zephyrproject
+         cd zephyrproject
+         west update
```
See comment about the pip3 sequence above

marc-hb reviewed (commented):
Not an very thorough review due to lack of time, apologies for any miss or noise.

dbkinder commented on doc/getting_started/index.rst:
```diff
@@ -3,319 +3,372 @@
+      .. code-block:: bash
 
-.. warning::
+         pip3 install -r https://raw.githubusercontent.com/zephyrproject-rtos/zephyr/master/scripts/requirements.txt
```
That's a nice solution for getting the requirements.txt locally for pip3 to use, but it does mess up the "Install tools" then "get source" flow (I think for the better because it eliminates the hard version reference).

nashif replied:
>  (I think for the better because it eliminates the hard version reference).

what version reference? are you referring to 'master'? This is just the same, it is just not implied.

In general I think we can solve this problem by introducing macros for releases and branches and avoid hardcoding the branch or the version so that such documentation has the right reference whether we are on master or on a release.

nashif replied:
> But I think that in general, content like this, as well as directives like `:zephyr_file:` suffer from the problem that they do not produce stable references to artifacts that are relevant to the content they are meant to document.

zephyr_file actually should work on branches, at least the implementation allows versioned references, maybe we are not using it correctly.

dbkinder replied:
Can’t use a reST substitution in the code block because that content isn’t processed. Sigh.

mbolivar replied:
> Can’t use a reST substitution in the code block because that content isn’t processed. Sigh.

You can use a parsed-literal if you're willing to give up syntax highlighting. Otherwise your only option is to write a Python extension that uses a variable set in conf.py to generate the code block nodes directly into the docutils DOM. But that's of course more complex.

mbolivar replied:
> what version reference? are you referring to 'master'? This is just the same, it is just not implied.

Right, it always refers to master, which becomes instantly wrong the moment the release is cut and the HTML stays the same while master changes.

> In general I think we can solve this problem by introducing macros for releases and branches and avoid hardcoding the branch or the version so that such documentation has the right reference whether we are on master or on a release.

Yes, but that has the problems @dbkinder pointed out. I've dealt with the same issue in the foundries documentation and the parsed-literal or extension options are the only ones I could find.

carlescufi replied:
I agree that pointing to the upstream `requirements.txt` is not going to work with documentation for previous releases, so it's better to always rely on the actual file on the disk post `west init`, because otherwise this will fail:

```
pip3 install -r https://raw.githubusercontent.com/zephyrproject-rtos/zephyr/master/scripts/requirements.txt
west init --mr v1.14.0
```
since the requirements of master will have been installed to clone v1.14.0.

mbolivar replied:
> west init --mr v1.14.0

That works for west init, but I just want to reiterate that this is a much more general concern about an issue throughout the documentation.

dbkinder commented on doc/getting_started/index.rst:
```diff
@@ -3,319 +3,372 @@
 
-.. toctree::
-   :maxdepth: 2
+      #. Set environment variables:
```
We've been trying to address issues via documentation, but it would also be great to address them with tool changes or additions too.

dbkinder commented on doc/getting_started/index.rst:
```diff
@@ -3,319 +3,372 @@
-.. tip::
-
-   Need help with something? See :ref:`help`.
+#. **Set up a command-line development environment** for Linux*, macOS*, or
```
We're not consistent about marking company trademarks in our documentation... :(

carlescufi replied:
It just looks strange or broken right now, so I would either remove it or fix it

dbkinder commented:
Thanks for the comments all.  I'm waiting for some additional feedback from a usability test using this new GSG and will be sending an update.

carlescufi commented on doc/getting_started/installation_linux.rst:
```diff
@@ -65,65 +70,77 @@ Install Requirements and Dependencies
+.. tabs::
 
-.. code-block:: console
+   .. group-tab:: Ubuntu
```
This doesn't seem to work? I only see the Ubuntu instructions, but no tabs with other operating systems

carlescufi replied:
Where is the Linux specific page in fact, I can't find it? I just see this:

![image](https://user-images.githubusercontent.com/12450381/65236508-0cdbca80-dad9-11e9-843e-f7725ca37fd3.png)

carlescufi replied:
Found it under the "Beyond the GSG" section, but as I write there, not very intuitive IMHO

carlescufi commented on doc/guides/beyond-GSG.rst:
```diff
@@ -0,0 +1,277 @@
+See `Installing Packages`_ in the Python Packaging User Guide for more
+information about pip\ [#pip]_, including `information on -\\-user`_.
+
+- On Linux, make sure ``~/.local/bin`` is at the front of your :envvar:`PATH`
```
The issue with this being in "beyond the GSG" is that this is required for many users in order to get a working build system. We already had complaints that this wasn't shown before the OS instructions, which is why we moved it to the top.

ru-fu replied:
But this part is in the GSG as well, and it is repeated as troubleshooting here - makes sense imo.

carlescufi commented on doc/guides/beyond-GSG.rst:
```diff
@@ -0,0 +1,277 @@
+.. _beyond-gsg:
+
+Beyond the Getting Started Guide
```
I don't think this should be under User Guides. This should instead be inside the GSG hierarchy in the ToC in my opinion.

carlescufi commented on doc/getting_started/index.rst:
```diff
@@ -3,319 +3,372 @@
+====================
 
-Follow an operating system specific guide, then come back to this page.
+Next, use your OS package manager to install required OS-provided
```
This is tricky because neither homebrew nor chocolatey are OS-provided package managers. They have to be installed on top of the OS as a separate application. Needs specifying what is what.

carlescufi commented on doc/getting_started/index.rst:
```diff
@@ -3,319 +3,372 @@
+            wget https://github.com/Kitware/CMake/releases/download/v3.15.3/cmake-3.15.3-Linux-x86_64.sh
+            chmod +x cmake-3.15.3-Linux-x86_64.sh
+            sudo ./cmake-3.15.3-Linux-x86_64.sh --skip-license --prefix=/usr/local
+            hash -r
```
Yes, in the Ubuntu section we could probably use those packages instead.

carlescufi commented on doc/getting_started/index.rst:
```diff
@@ -3,319 +3,372 @@
 
-Install Python Dependencies
-***************************
+Install Zephyr Software Development Kit (SDK)
```
Agreed. Or, alternatively, change the name to "Installing a toolchain" and use the Zephyr SDK instrucitons in the Linux tab and the instructions for 3rd-party toolchains in the macOS and Windows tabs

carlescufi reviewed (dismissed):
Looks pretty nice! But I think we need some changes to make this work properly.

ru-fu commented on doc/getting_started/index.rst:
```diff
@@ -3,319 +3,372 @@
+      An easy way to install native Windows dependencies is to first install
+      `Chocolatey`_, a package manager for Windows.  If you prefer to install
+      dependencies manually, you can also download the required programs from their
+      respective websites and verify they can be found on your PATH.
```
Does chocolatey add the tools to the PATH? If I remember correctly, it just gives you an output of what to add to the PATH. In that case, we should add a step as a reminder to actually do that.

ru-fu commented on doc/getting_started/index.rst:
```diff
@@ -3,319 +3,372 @@
-on a POSIX OS. This is currently only tested on Linux hosts. See
-:ref:`native_posix` for more information. On 64 bit host operating systems, you
-need to install a 32 bit C library; see :ref:`native_posix_deps` for details.
+#. Build the blinky sample. Specify **your board** in the command below.
```
Maybe add a link to the supported boards?

ru-fu commented on doc/getting_started/index.rst:
```diff
@@ -3,319 +3,372 @@
-.. _Installing Packages: https://packaging.python.org/tutorials/installing-packages/
+* Try some other :ref:`samples-and-demos` that demonstrate Zephyr
+  capabilities.
+* Learn about :ref:`application` and more details about :ref:`west`
```
Period at the end (also two lines down).

ru-fu commented on doc/getting_started/installation_mac.rst:
```diff
@@ -3,10 +3,6 @@
 Install macOS Host Dependencies
```
This looks more like a troubleshooting section than "alternative instructions". It's duplication of the content in Getting Started plus some more information - maybe these sections should focus on troubleshooting?

ru-fu commented on doc/getting_started/installation_win.rst:
```diff
@@ -3,10 +3,6 @@
 Install Windows Host Dependencies
```
As for macOS, this is more a repetition plus troubleshooting. Maybe turn it into a pure troubleshooting section?

ru-fu commented on doc/guides/beyond-GSG.rst:
```diff
@@ -0,0 +1,277 @@
+and build project documentation.
+
+Depending on your operating system, you may need to provide the
+``--user`` flag to the ``pip3`` command when installing new packages. This is
```
We could be clearer here and explicitly state that you need the --user for Linux.

marc-hb replied:
No because it depends which "Linux".

ru-fu replied:
.. for all supported operating systems except for macOS and Windows? ;)
At least that's how it's currently documented.

marc-hb replied:
From: https://docs.brew.sh/Homebrew-and-Python#setuptools-pip-etc
> The normal pip install --user is disabled for brewed Python. This is because of a bug in distutils

So this could change on macOS any time after  that bug is fixed.

On Debian and most other debian-based distros (including Ubuntu?):
https://github.com/pypa/pip/issues/4222#issuecomment-417646535
> When running as a normal user in a non-virtual environment, default to
--user and --ignore-installed.  When inside virtual environments or when
running as root, keep the default behavior.

Not the case on other Linux distributions.

It's a mess so there's no way to clarify the "Depending on your operating system" part, you can stop trying. The only solution is to:
1. focus on and try to stay up to date with a *limited* number of operating systems: Ubuntu, macOS, Windows, Period.
2. defer to OS-specific documentations for other (Linux) operating systems

ru-fu commented on doc/guides/beyond-GSG.rst:
```diff
@@ -0,0 +1,277 @@
+See `Installing Packages`_ in the Python Packaging User Guide for more
+information about pip\ [#pip]_, including `information on -\\-user`_.
+
+- On Linux, make sure ``~/.local/bin`` is at the front of your :envvar:`PATH`
```
Should these be tabs as in the GSG?

ru-fu commented on doc/guides/beyond-GSG.rst:
```diff
@@ -0,0 +1,277 @@
+Advanced Setup and tool chain alternatives
+******************************************
+
+Here are some alternative instructions for more advanced platform setup
```
Need to decide what we want to have here - repetition of the GSG, alternative setups, or troubleshooting. Currently, it's a bit of a mix.

ru-fu commented on doc/guides/beyond-GSG.rst:
```diff
@@ -0,0 +1,277 @@
+
+.. _gs_toolchain:
+
+Set Up a Toolchain
```
We should have a link to this section from the GSG, because you need to install one (unless you're on Linux and installed the SDK).

ru-fu commented on doc/guides/beyond-GSG.rst:
```diff
@@ -0,0 +1,277 @@
+   ../getting_started/toolchain_host.rst
+   ../getting_started/toolchain_custom_cmake.rst
+
+Cloning the Zephyr Repositories
```
I like the idea of giving more detail in this guide than in the GSG, but it must be made clearer what is repetition and what is additional content. Also, we should have links from each section in the GSG to the "more information if you need it" section here.

dbkinder commented:
I've incorporated changes based on your comments and additional DX usability testing.  Overall, this updated and simplified GSG approach has significantly improved the initial setup experience of Zephyr.  When the SDK improvements are released we can submit a new PR to update these instructions with some further simplifications and time improvements, followed by more DX improvements over time.

dbkinder commented:
And you can view the [build artifacts for this PR](https://builds.zephyrproject.org/zephyrproject-rtos/zephyr/19123/docs/getting_started/index.html) to see the generated HTML, much easier than staring at the reST files :)

andrewboie commented on doc/getting_started/installation_linux.rst:
```diff
@@ -169,6 +198,8 @@ following target architectures:
 
 * :abbr:`X86 (Intel Architecture 32 bits)`
 
+* :abbr:`X86 IAMCU ABI (Intel Architecture 32 bits IAMCU ABI)`
```
IAMCU has been purged from Zephyr

mbolivar commented on doc/getting_started/index.rst:
```diff
@@ -3,319 +3,409 @@
+
+      #. Install west:
+
+         .. code-block:: bash
```
Please use the 'bat' lexer (not bash) for windows cmd.exe syntax highlighting. Some of these lines look weird rendered for bash.

mbolivar commented on doc/getting_started/index.rst:
```diff
@@ -3,319 +3,409 @@
-Run the Application in QEMU
-===========================
+            cd ~/zephyrproject/zephyr
+            source zephyr-env.sh
```
I'd still like to eliminate explicit use of environment variables in the command line flow using west someday.

Would you be open to removing these scripts from the GSG at that time?

dbkinder replied:
Yes, if we can assume that EVERYONE will be using west :)

mbolivar replied:
> Yes, if we can assume that EVERYONE will be using west :)

We can keep environment variable support around as an advanced topic or for people who don't use west, just talking about the GSG and 'recommended easiest' workflows here.

mbolivar commented on doc/getting_started/index.rst:
```diff
@@ -3,319 +3,409 @@
+   .. code-block:: bash
 
-.. _gs_posix:
+      west build -p -b <your-board-name> samples/basic/blinky
```
Mixed feelings about using --pristine here, as it instills a "bad" habit of never doing incremental builds.

But not asking for a change, just pointing this out.

dbkinder replied:
Yes, I know that.  But for this simplified instruction path it was an easy way to eliminate problems seen by test subjects that chose the wrong board and then tried specifying the right board without cleaning out the build folder (bad things happen if you change the board or application).  I could change this to ``-p auto`` I suppose...

mbolivar replied:
Right, I'm not asking for a change, but maybe this means 'doing incremental builds' ought to be a prominent 'beyond the GSG' topic.

mbolivar commented on doc/getting_started/installation_linux.rst:
```diff
@@ -19,35 +15,44 @@ requirements and dependencies may not be met by your package manager. In that
+   need to configure a proxy for accessing the internet, if you haven't
+   done so already.  While some tools use the environment variables
+   ``http_proxy`` and ``https_proxy`` to get their proxy settings, some
+   use their own configuration files, most notibly ``apt`` and
```
```suggestion
   use their own configuration files, most notably ``apt`` and
```

mbolivar commented on doc/getting_started/installation_linux.rst:
```diff
@@ -65,65 +70,77 @@ Install Requirements and Dependencies
+      .. code-block:: console
 
-Fedora:
+         sudo apt-get install --no-install-recommends git cmake ninja-build gperf \
```
This also will need updates if you remove the extra apt install items from the GSG itself.

mbolivar commented on doc/getting_started/installation_linux.rst:
```diff
@@ -65,65 +70,77 @@ Install Requirements and Dependencies
 
-     pip3 install --user cmake
+     cd ~
+     wget https://github.com/Kitware/CMake/releases/download/v3.15.3/cmake-3.15.3-Linux-x86_64.sh
```
Please don't make everyone else install an entirely separate cmake into ~/.local just because Ubuntu 18.04 and 18.10 have outdated versions.

mbolivar commented on doc/getting_started/installation_linux.rst:
```diff
@@ -65,65 +70,77 @@ Install Requirements and Dependencies
+     cd ~
+     wget https://github.com/Kitware/CMake/releases/download/v3.15.3/cmake-3.15.3-Linux-x86_64.sh
+     chmod +x cmake-3.15.3-Linux-x86_64.sh
+     sudo ./cmake-3.15.3-Linux-x86_64.sh --skip-license --prefix=/usr/local
```
I think some comments both here and in the GSG saying "go update the other file if you change this" would ease maintainability.

mbolivar commented on doc/guides/beyond-GSG.rst:
```diff
@@ -0,0 +1,277 @@
+.. _beyond-gsg:
```
Haven't had time to review this one yet, but I'm excited about it!

mbolivar reviewed (dismissed):
Very nice update. Some mostly minor things.

mbolivar commented on doc/getting_started/index.rst:
```diff
@@ -3,319 +3,409 @@
+            sudo apt install --no-install-recommends git cmake ninja-build gperf \
+              ccache dfu-util device-tree-compiler wget \
+              python3-pip python3-setuptools python3-tk python3-wheel xz-utils file \
+              make gcc gcc-multilib
```
Are some of those perhaps for native_posix? Maybe they can be removed from GSG and into beyond-the-GSG if so?

dbkinder replied:
I don't think the extra packages are a problem (for now) since we're not quite sure if there are dependencies for STMicro or native_posix builds.  We can remove them later if/when we know.

dbkinder commented:
@mbolivar @nashif @andrewboie @ru-fu @carlescufi @aescolar @marc-hb I'm hoping this PR is close enough to get merged as an improvement to the existing GSG and we can continue with other DX improvements (and improvements to this documentation) in subsequent PRs...

marc-hb reviewed (commented):
This PR is pretty big and github is not able to show the difference force-pushed since a last review. I'm out, sorry.

mbolivar reviewed (changes requested):
-1 just from a quick check that sudo make flash is still there; haven't had time to look at the new batch of changes otherwise

dleach02 commented:
@dbkinder , What if we consider a reference link to a new page on udev needs? We can then add the appropriate udev rules as they are identified for the various target boards. That would than remove the desire to have the sudo command for west flash

andrewboie commented:
Worth pointing out: if you clone the openocd-git repo:

```
git clone https://git.code.sf.net/p/openocd/code openocd-code
```

There's a file contrib/60-openocd.rules which at least for me has allowed me to flash every Zephyr supported board I have tried so far. There might not be support for *every* board, but at least it gets you pretty far:

```
$ cat contrib/60-openocd.rules 
# Copy this file to /etc/udev/rules.d/
# If rules fail to reload automatically, you can refresh udev rules
# with the command "udevadm control --reload"

ACTION!="add|change", GOTO="openocd_rules_end"
SUBSYSTEM!="usb|tty|hidraw", GOTO="openocd_rules_end"

# Please keep this list sorted by VID:PID

# opendous and estick
ATTRS{idVendor}=="03eb", ATTRS{idProduct}=="204f", MODE="660", GROUP="plugdev", TAG+="uaccess"

# Original FT232/FT245 VID:PID
ATTRS{idVendor}=="0403", ATTRS{idProduct}=="6001", MODE="660", GROUP="plugdev", TAG+="uaccess"

# Original FT2232 VID:PID
ATTRS{idVendor}=="0403", ATTRS{idProduct}=="6010", MODE="660", GROUP="plugdev", TAG+="uaccess"

# Original FT4232 VID:PID
ATTRS{idVendor}=="0403", ATTRS{idProduct}=="6011", MODE="660", GROUP="plugdev", TAG+="uaccess"

# Original FT232H VID:PID
ATTRS{idVendor}=="0403", ATTRS{idProduct}=="6014", MODE="660", GROUP="plugdev", TAG+="uaccess"

# DISTORTEC JTAG-lock-pick Tiny 2
ATTRS{idVendor}=="0403", ATTRS{idProduct}=="8220", MODE="660", GROUP="plugdev", TAG+="uaccess"

# TUMPA, TUMPA Lite
ATTRS{idVendor}=="0403", ATTRS{idProduct}=="8a98", MODE="660", GROUP="plugdev", TAG+="uaccess"
ATTRS{idVendor}=="0403", ATTRS{idProduct}=="8a99", MODE="660", GROUP="plugdev", TAG+="uaccess"

# XDS100v2
ATTRS{idVendor}=="0403", ATTRS{idProduct}=="a6d0", MODE="660", GROUP="plugdev", TAG+="uaccess"

# Xverve Signalyzer Tool (DT-USB-ST), Signalyzer LITE (DT-USB-SLITE)
ATTRS{idVendor}=="0403", ATTRS{idProduct}=="bca0", MODE="660", GROUP="plugdev", TAG+="uaccess"
ATTRS{idVendor}=="0403", ATTRS{idProduct}=="bca1", MODE="660", GROUP="plugdev", TAG+="uaccess"

# TI/Luminary Stellaris Evaluation Board FTDI (several)
ATTRS{idVendor}=="0403", ATTRS{idProduct}=="bcd9", MODE="660", GROUP="plugdev", TAG+="uaccess"

# TI/Luminary Stellaris In-Circuit Debug Interface FTDI (ICDI) Board
ATTRS{idVendor}=="0403", ATTRS{idProduct}=="bcda", MODE="660", GROUP="plugdev", TAG+="uaccess"

# egnite Turtelizer 2
ATTRS{idVendor}=="0403", ATTRS{idProduct}=="bdc8", MODE="660", GROUP="plugdev", TAG+="uaccess"

# Section5 ICEbear
ATTRS{idVendor}=="0403", ATTRS{idProduct}=="c140", MODE="660", GROUP="plugdev", TAG+="uaccess"
ATTRS{idVendor}=="0403", ATTRS{idProduct}=="c141", MODE="660", GROUP="plugdev", TAG+="uaccess"

# Amontec JTAGkey and JTAGkey-tiny
ATTRS{idVendor}=="0403", ATTRS{idProduct}=="cff8", MODE="660", GROUP="plugdev", TAG+="uaccess"

# TI ICDI
ATTRS{idVendor}=="0451", ATTRS{idProduct}=="c32a", MODE="660", GROUP="plugdev", TAG+="uaccess"

# STMicroelectronics ST-LINK V1
ATTRS{idVendor}=="0483", ATTRS{idProduct}=="3744", MODE="660", GROUP="plugdev", TAG+="uaccess"

# STMicroelectronics ST-LINK/V2
ATTRS{idVendor}=="0483", ATTRS{idProduct}=="3748", MODE="660", GROUP="plugdev", TAG+="uaccess"

# STMicroelectronics ST-LINK/V2.1
ATTRS{idVendor}=="0483", ATTRS{idProduct}=="374b", MODE="660", GROUP="plugdev", TAG+="uaccess"
ATTRS{idVendor}=="0483", ATTRS{idProduct}=="3752", MODE="660", GROUP="plugdev", TAG+="uaccess"

# STMicroelectronics STLINK-V3
ATTRS{idVendor}=="0483", ATTRS{idProduct}=="374d", MODE="660", GROUP="plugdev", TAG+="uaccess"
ATTRS{idVendor}=="0483", ATTRS{idProduct}=="374e", MODE="660", GROUP="plugdev", TAG+="uaccess"
ATTRS{idVendor}=="0483", ATTRS{idProduct}=="374f", MODE="660", GROUP="plugdev", TAG+="uaccess"
ATTRS{idVendor}=="0483", ATTRS{idProduct}=="3753", MODE="660", GROUP="plugdev", TAG+="uaccess"

# Cypress KitProg in KitProg mode
ATTRS{idVendor}=="04b4", ATTRS{idProduct}=="f139", MODE="660", GROUP="plugdev", TAG+="uaccess"

# Cypress KitProg in CMSIS-DAP mode
ATTRS{idVendor}=="04b4", ATTRS{idProduct}=="f138", MODE="660", GROUP="plugdev", TAG+="uaccess"

# Hilscher NXHX Boards
ATTRS{idVendor}=="0640", ATTRS{idProduct}=="0028", MODE="660", GROUP="plugdev", TAG+="uaccess"

# Hitex STR9-comStick
ATTRS{idVendor}=="0640", ATTRS{idProduct}=="002c", MODE="660", GROUP="plugdev", TAG+="uaccess"

# Hitex STM32-PerformanceStick
ATTRS{idVendor}=="0640", ATTRS{idProduct}=="002d", MODE="660", GROUP="plugdev", TAG+="uaccess"

# Altera USB Blaster
ATTRS{idVendor}=="09fb", ATTRS{idProduct}=="6001", MODE="660", GROUP="plugdev", TAG+="uaccess"

# Amontec JTAGkey-HiSpeed
ATTRS{idVendor}=="0fbb", ATTRS{idProduct}=="1000", MODE="660", GROUP="plugdev", TAG+="uaccess"

# SEGGER J-Link
ATTRS{idVendor}=="1366", ATTRS{idProduct}=="0101", MODE="660", GROUP="plugdev", TAG+="uaccess"
ATTRS{idVendor}=="1366", ATTRS{idProduct}=="0102", MODE="660", GROUP="plugdev", TAG+="uaccess"
ATTRS{idVendor}=="1366", ATTRS{idProduct}=="0103", MODE="660", GROUP="plugdev", TAG+="uaccess"
ATTRS{idVendor}=="1366", ATTRS{idProduct}=="0104", MODE="660", GROUP="plugdev", TAG+="uaccess"
ATTRS{idVendor}=="1366", ATTRS{idProduct}=="0105", MODE="660", GROUP="plugdev", TAG+="uaccess"
ATTRS{idVendor}=="1366", ATTRS{idProduct}=="0107", MODE="660", GROUP="plugdev", TAG+="uaccess"
ATTRS{idVendor}=="1366", ATTRS{idProduct}=="0108", MODE="660", GROUP="plugdev", TAG+="uaccess"
ATTRS{idVendor}=="1366", ATTRS{idProduct}=="1010", MODE="660", GROUP="plugdev", TAG+="uaccess"
ATTRS{idVendor}=="1366", ATTRS{idProduct}=="1011", MODE="660", GROUP="plugdev", TAG+="uaccess"
ATTRS{idVendor}=="1366", ATTRS{idProduct}=="1012", MODE="660", GROUP="plugdev", TAG+="uaccess"
ATTRS{idVendor}=="1366", ATTRS{idProduct}=="1013", MODE="660", GROUP="plugdev", TAG+="uaccess"
ATTRS{idVendor}=="1366", ATTRS{idProduct}=="1014", MODE="660", GROUP="plugdev", TAG+="uaccess"
ATTRS{idVendor}=="1366", ATTRS{idProduct}=="1015", MODE="660", GROUP="plugdev", TAG+="uaccess"
ATTRS{idVendor}=="1366", ATTRS{idProduct}=="1016", MODE="660", GROUP="plugdev", TAG+="uaccess"
ATTRS{idVendor}=="1366", ATTRS{idProduct}=="1017", MODE="660", GROUP="plugdev", TAG+="uaccess"
ATTRS{idVendor}=="1366", ATTRS{idProduct}=="1018", MODE="660", GROUP="plugdev", TAG+="uaccess"

# Raisonance RLink
ATTRS{idVendor}=="138e", ATTRS{idProduct}=="9000", MODE="660", GROUP="plugdev", TAG+="uaccess"

# Debug Board for Neo1973
ATTRS{idVendor}=="1457", ATTRS{idProduct}=="5118", MODE="660", GROUP="plugdev", TAG+="uaccess"

# Olimex ARM-USB-OCD
ATTRS{idVendor}=="15ba", ATTRS{idProduct}=="0003", MODE="660", GROUP="plugdev", TAG+="uaccess"

# Olimex ARM-USB-OCD-TINY
ATTRS{idVendor}=="15ba", ATTRS{idProduct}=="0004", MODE="660", GROUP="plugdev", TAG+="uaccess"

# Olimex ARM-JTAG-EW
ATTRS{idVendor}=="15ba", ATTRS{idProduct}=="001e", MODE="660", GROUP="plugdev", TAG+="uaccess"

# Olimex ARM-USB-OCD-TINY-H
ATTRS{idVendor}=="15ba", ATTRS{idProduct}=="002a", MODE="660", GROUP="plugdev", TAG+="uaccess"

# Olimex ARM-USB-OCD-H
ATTRS{idVendor}=="15ba", ATTRS{idProduct}=="002b", MODE="660", GROUP="plugdev", TAG+="uaccess"

# USBprog with OpenOCD firmware
ATTRS{idVendor}=="1781", ATTRS{idProduct}=="0c63", MODE="660", GROUP="plugdev", TAG+="uaccess"

# TI/Luminary Stellaris In-Circuit Debug Interface (ICDI) Board
ATTRS{idVendor}=="1cbe", ATTRS{idProduct}=="00fd", MODE="660", GROUP="plugdev", TAG+="uaccess"

# TI XDS110 Debug Probe (Launchpads and Standalone)
ATTRS{idVendor}=="0451", ATTRS{idProduct}=="bef3", MODE="660", GROUP="plugdev", TAG+="uaccess"

# TI Tiva-based ICDI and XDS110 probes in DFU mode
ATTRS{idVendor}=="1cbe", ATTRS{idProduct}=="00ff", MODE="660", GROUP="plugdev", TAG+="uaccess"

# Ambiq Micro EVK and Debug boards.
ATTRS{idVendor}=="2aec", ATTRS{idProduct}=="6010", MODE="664", GROUP="plugdev", TAG+="uaccess"
ATTRS{idVendor}=="2aec", ATTRS{idProduct}=="6011", MODE="664", GROUP="plugdev", TAG+="uaccess"
ATTRS{idVendor}=="2aec", ATTRS{idProduct}=="1106", MODE="664", GROUP="plugdev", TAG+="uaccess"

# Marvell Sheevaplug
ATTRS{idVendor}=="9e88", ATTRS{idProduct}=="9e8f", MODE="660", GROUP="plugdev", TAG+="uaccess"

# Keil Software, Inc. ULink
ATTRS{idVendor}=="c251", ATTRS{idProduct}=="2710", MODE="660", GROUP="plugdev", TAG+="uaccess"

# CMSIS-DAP compatible adapters
ATTRS{product}=="*CMSIS-DAP*", MODE="660", GROUP="plugdev", TAG+="uaccess"

LABEL="openocd_rules_end"
```

Ideally though, the documentation for every board contains information on the necessary udev string to get it going.

dbkinder commented:
> Ideally though, the documentation for every board contains information on the necessary udev string to get it going.

Alas, a quick look shows that only 4 boards (out of over 200) mention udev rules in their documentation.  Not sure I'm ready to submit PR(s) to touch all the board docs, so maybe indeed the notion of a udev troubleshooting section in the linux-specific doc is the answer.  (And have ``west flash`` mention udev rules if it has trouble flashing.)

marc-hb commented:
> so maybe indeed the notion of a udev troubleshooting section in the linux-specific doc is the answer. 

For some strange reason, everyone on the internet and their dog recommends `udevadm test` to troubleshoot what `udevd` does.  But `udevadm test` provides zero log or trace, it's only a simulator. Tentative fix: https://superuser.com/questions/677106/how-to-check-if-a-udev-rule-fired/1501568#1501568

dbkinder commented:
Flashing with sudo is out now, use of the openOCD udev rules file is in.  @marc-hb sorry about the forced pushes hiding what changes were made. :(

mbolivar commented on doc/getting_started/index.rst:
```diff
@@ -3,319 +3,445 @@
+         Add the kitware repo corresponding to your Ubuntu release (use
+         ``cat /etc/os-release`` to check):
+
+         - For Ubuntu Bionic Beaver (18.04) use:
```
@dbkinder Ubuntu 18.10 ships CMake 3.12, so these instructions won't work for that version of Ubuntu, and the user will still have an out of date CMake.

Ubuntu 19.04 (and thus all subsequent versions) ship a recent enough version.

To fix these instructions for 18.10, you could just add a generic reference to their download page:

https://cmake.org/download/

andrewboie replied:
Maybe I'm missing something here, but since CMake is available with pip, couldn't we just add cmake to scripts/requirements.txt and just standardize on that for every host OS?

mbolivar commented on doc/getting_started/index.rst:
```diff
@@ -3,319 +3,445 @@
+
+              sudo apt-add-repository 'deb https://apt.kitware.com/ubuntu/ bionic main'
+
+         - For Ubuntu Xeniel Xerus (16.04) use:
```
```suggestion
         - For Ubuntu Xenial Xerus (16.04) use:
```

mbolivar commented:
> There's a file contrib/60-openocd.rules which at least for me has allowed me to flash every Zephyr supported board I have tried so far. There might not be support for _every_ board, but at least it gets you pretty far:

There's an openocd rules file in the SDK, FWIW.

Installing that file into /etc/udev/rules.d only works if your user is in the plugdev group, of course. And then you have to reset the udev state machines or unplug/replug in your device.

Fiddling with udev rules is a very annoying rite of passage doing embedded development on Linux in my experience, unfortunately. I blame greg KH :).

marc-hb commented:
> @marc-hb sorry about the forced pushes hiding what changes were made. :(

At the high risk of stating the obvious: this is a github limitation, nothing about you (or any other user) github's preferred review model is based on "fixup" commits and squash. So it's compatible with neither force pushes nor the review of a series of multiple commits. Off-topic sorry, in the unlikely urge to discuss further use #14444 or the #github-usage channel on Slack

mbolivar reviewed (approved):
Looks great! Thanks for being willing to do extra work to help users troubleshoot udev rules, and for the awesome new GSG!

mbolivar commented:
@dbkinder thanks for rounding up some of the discussion on udev rules into the force-push you just made. That wasn't necessary IMO, but I appreciate it!

This comment could eventually be added too, if you want to improve this further:

> Installing that file into /etc/udev/rules.d only works if your user is in the plugdev group

According to https://wiki.ubuntu.com/Security/Privileges, only the first user added to an Ubuntu desktop system is a member of this group. So multi-user systems (admittedly very rare) could run into issues here.

dbkinder commented:
Thanks @mbolivar I think I'll save that for another update.  Time to get this one merged :)

mbolivar commented:
> Thanks @mbolivar I think I'll save that for another update. Time to get this one merged :)

Yep yep, not asking for a change, just making a note.

andrewboie commented:
something I think is worth pointing out: you can install a suitable cmake on Ubuntu using snap. This seems a lot simpler than the current instructions.

```
sudo snap install cmake
```

dbkinder commented:
@andrewboie It is listed as an alternative in the linux-specific detailed doc (https://builds.zephyrproject.org/zephyrproject-rtos/zephyr/19123/docs/getting_started/installation_linux.html#cmake) along with using pip :)

I can submit an update after this big and long-delayed PR gets merged...

dbkinder commented:
> something I think is worth pointing out: you can install a suitable cmake on Ubuntu using snap. This seems a lot simpler than the current instructions.
> 
> ```
> sudo snap install cmake
> ```

I gave it a try and needed to configure the snap proxy and use the ``--classic`` option
```
sudo snap set system proxy.http=...
sudo snap set system proxy.https=...
sudo snap install cmake --classic
```
Snap puts its binaries into ``/snap/bin`` so need to add that to the PATH (not unlike the problems users have with ``pip --user`` and not adding ``~/.local/bin``.  Sigh.
